### ISSUE TITLE
Make wchoose and windex use relative weights (normalize by default)

### ISSUE THREAD

**Issue by Asmatzaile** [supercollider/supercollider#6311]
## Motivation

`wchoose` and `windex` don't normalize; that makes it easy to get unexpected results.

```
(
var list, pick;
list= [\red, \green, \blue];
pick = list.wchoose([100,200,300]);
)
```

Looking at the above code, one would expect to get `\blue` thrice as often and `\green` twice as often as `\red`. In reality,`\red` is the only element that is picked.

The correct option would have been to use `list.wchoose([100,100,100].normalizeSum)`. In practice, all lists with weights that are not normalized (and I assume that they are the majority) will need to have `.normalizeSum` applied.

## Description of Proposed Feature

The proposal is to make those functions normalize by default, as other languages as [python](https://docs.python.org/3/library/random.html#random.choices). If efficiency is a concern, a flag can be added for the arguably rare cases where the sum of the list is already equal to one.

**Comment by smoge** [comment]
For reference, one of the common modules for this in Haskell is also normalized by default

 \( \frac{w}{\text{sum of all weights}} \)

>  
>  Weighted list
>     A list of pairs (w, a), where w is the weight of element a. The probability of an element getting into the first position is equal by its weight divided by the sum of all weights, and the probability of getting into a position other than the first is equal to the probability of getting in the first position when all elements in prior positions have been removed from the weighted list. 

 https://hackage.haskell.org/package/random-extras-0.19/docs/Data-Random-Shuffle-Weighted.html

**Comment by scztt** [comment]
The documentation states that weights *must* normalize to 1.0. The behavior if they do not is not defined in the documentation and not exactly expected (final item(s) in the list either have much higher probability or dont occur at all, depending on the sum of the weights).

As it is, user code must either (a) call `.normalize` on weights, or (b) use weights that naturally sum to 1.0 - almost all does (a). Performance will be *slightly* slower for old code in either of these cases, but the output will not change. However, new code doesn't need `.normalize`, and normalizing internally will be much faster because it can be done in place and doesn't require a full array copy.

In short: we eliminate the possibility of non-obvious invalid behavior and make this faster for new code.

**Comment by smoge** [comment]
PR material

**Comment by smoge** [comment]
@scztt 
I thought of some cases where nobody loses performance. Does anyone want the honors to submit a first version as PR, or it will have to be me? Do you want to review, maybe? It would be cool.

**Comment by smoge** [comment]
I thought of some cases where nobody loses performance. Does anyone want the honor of submitting the first draft, or should I go ahead?

I don't care how does it, I'm just volunteering.

Would you like to be one of the reviewers, @scztt? That would be cool.

cheers

**Comment by Asmatzaile** [comment]
@smoge go ahead for me! I don't know how to
> normaliz[e] internally [so that it] doesn't require a full array copy

**Comment by telephon** [comment]
While I really like the idea, I wonder how we can make this efficient:

```
(
var n = 2048;
var weights = Array.fill(n, { |i| i.squared.rand }).normalizeSum;
var freqs = Array.fill(n, { exprand(300, 8000) });
fork {
     loop {
          0.01.wait;
          (freq: freqs.wchoose(weights), sustain: 0.01).play;
     }
}
)
```

**Comment by smoge** [comment]
> While I really like the idea, I wonder how we can make this efficient for high n:

`windex` is already a primitive, if performance is the priority, this could be an option. It is  linear in complexity $O(n)$  

https://en.cppreference.com/w/cpp/numeric/random/discrete_distribution

**Comment by telephon** [comment]
The bottleneck is not `wchoose` but `normalizeSum`. If you normalize for each single random value, this is an unnecessary overhead, and if you pass a normalized list, it is a wasted effort.

I would say it is better to make an extra method `wchooseN` which normalizes.

```supercollider
wchooseN { |weights| ^this.wchoose(weights.normalizeSum) }
```

**Comment by smoge** [comment]
> The bottleneck is not `wchoose` but `normalizeSum`. If you normalize for each single random value, this is an unnecessary overhead, and if you pass a normalized list, it is a wasted effort and may even result in (small) floating point differences.

If you just check something like `weights.sum.equalWithPrecision(1.0, 0.001)` (to deal with floating-point errors) in an if statement, would that make it? A normalized list would not be normalized again.

**Comment by smoge** [comment]
I did some benchmarks comparing a regular algorithm and another one using a parallel implementation. It got worse with small lists, so I made a hybrid that would select the parallel or regular implementation based on the list size. 

But for SuperCollider common use cases, it does not help, the benefits only start to show up with lists around 30.000 and larger.

```
-------------------------------------------------------------------------
Benchmark                               Time             CPU   Iterations
-------------------------------------------------------------------------
BM_NormalizeListHybrid/1024         41870 ns        41767 ns        16663
BM_NormalizeListHybrid/4096        166876 ns       166408 ns         4188
BM_NormalizeListHybrid/32768       710370 ns       488787 ns         1445
BM_NormalizeListHybrid/262144     1772666 ns       503234 ns         1391
BM_NormalizeListRegular/1024        41502 ns        41288 ns        16813
BM_NormalizeListRegular/4096       158285 ns       157763 ns         4228
BM_NormalizeListRegular/32768     1277059 ns      1272980 ns          530
BM_NormalizeListRegular/262144   10478278 ns     10445453 ns           66
```

**Comment by smoge** [comment]
Testing normalizeSum in cpp with `std:accumulation` to sum large lists, I got good results.

`std::accumulate`  is analogous to `foldl` in Haskell


https://gist.github.com/smoge/e39070de46007681f47b1e313312e2a9

//  g++ -std=c++20 -o sum6  sum6.cc -O2

// Size: 1000
// Sum after normalization: 1
// delta time: 2.0869e-05s

// Size: 10000
// Sum after normalization: 1
// delta time: 0.000135792s

// Size: 100000
// Sum after normalization: 1
// delta time: 0.00131085s

// Size: 1000000
// Sum after normalization: 1
// delta time: 0.0146801s

**Comment by telephon** [comment]
Yes, good as an improvement of normalizeSum – it just requires that the array consists of numbers only. Maybe one could do it for FloatAttay, or bail out if a non-number suddenly turns up in the list and then do it in sclang instead.

For the current issue, I would say that it is better to go with an extra pair of methods.

But if you like, it would be good to have an extra one for an optimised normalizedSum. It is a classical candidate for this, only the fact that normal arrays may have different objects is a minor obstacle to care for.

**Comment by smoge** [comment]
Checking if all items are numbers, still gives good results. But it would have to be tested inside sclang to be sure. I don't know what algorithm FloatArray uses, but maybe it can be an improvement there too.

It's just an idea. Let me know where it will be more helpful. 

Benchmark with a check if item is a number:

Size: 1000
Sum before normalization: 50785.3
Sum after normalization: 1
Elapsed time: 2.177e-05s

Size: 10000
Sum before normalization: 495409
Sum after normalization: 1
Elapsed time: 0.000132373s

Size: 100000
Sum before normalization: 5.01073e+06
Sum after normalization: 1
Elapsed time: 0.0012734s

Size: 1000000
Sum before normalization: 5.00209e+07
Sum after normalization: 1
Elapsed time: 0.0144793s

**Comment by telephon** [comment]
> Broader question: From a language design perspective, what would mean a string in an Array.normalizeSum? Does it work right now?

well, generally, `normalizeSum = [x_1 / (x_1 + x_2 + ...), x2 / (x_1 + x_2 + ...), ... ]`. So that we have two things:
- the sum of the elements
- the division of each element by the sum

What would be the sum of the elements of a string? Characters can't be added, but suppose they could, they may form a string
Then what could it mean to divide a character by a string?

For other things, the above make sense more easily, for example for UGens.

**Comment by smoge** [comment]
windex checks if all items are numbers:

```
    for (i = 0; i < size; ++i) {
        err = getIndexedDouble(obj, i, &w);
        if (err)
            return err;
        sum += w;
        if (sum >= r) {
            j = i;
            break;
        }
    }
```

There is a primitive prArrayNormalizeSum, is it just used for FloatArray? 

I'm just surprised because I've never seen .normalizeSum being used in other contexts...

**Comment by telephon** [comment]
Ah yes – I should have read the code. It already does more or less what I thought it should do.

>I'm just surprised because I've never seen .normalizeSum being used in other contexts...

Well, I suppose the point is never what we have seen, but what is made possible.

But `windex` currently throws errors and is rather "conservative". We use `TWindex` where UGens are concerned, which is more efficient than directly building the graph.

**Comment by smoge** [comment]
Ok, so I will save the idea to use in a new method for Array and/or FloatArray. One possible name: `normalizeVectorSum` (better than `normalizeSum2`, right?)

```
void normalize_array(std::vector<double>& arr) {
  const double sum = std::accumulate(arr.begin(), arr.end(), 0.0);
  if (sum <= std::numeric_limits<double>::epsilon()) {
    throw std::runtime_error("Cannot normalize array");
  }
  double inv_sum = 1.0 / sum;
  std::transform(arr.begin(), arr.end(), arr.begin(), [inv_sum](double val) { return val * inv_sum; });
}
```


// Size: 1000
// Generation time: 8.925e-06s
// Normalization time: 1.593e-06s
// Size: 10000
// Generation time: 8.8198e-05s
// Normalization time: 1.5404e-05s
// Size: 100000
// Generation time: 0.000832365s
// Normalization time: 0.000155685s
// Size: 1000000
// Generation time: 0.00833432s
// Normalization time: 0.00181367s

**Comment by telephon** [comment]
So @Asmatzaile – would you like to reopen another issue or rename this one? We should not propose to chenge windex and wchoose, but instead add two new methods.

**Comment by smoge** [comment]
> Is checking that the sum of an array is 1 so expensive?

In SCLang, checking this would be quite expensive.

**Comment by telephon** [comment]
My issue is that it takes away a possibility of efficiently calling wchoose on a huge array (or at least that it is a waste of energy). This is why I propesed separate methods. 

you could maybe benchmark with this?

```supercollider
(
var n = 10000;
var weights = Array.fill(n, { |i| i.squared.rand }).normalizeSum;
var freqs = Array.fill(n, { exprand(300, 8000) });
bench { 1000.do { freqs.wchoose(weights) } };
)

(
var n = 10000;
var weights = Array.fill(n, { |i| i.squared.rand });
var freqs = Array.fill(n, { exprand(300, 8000) });
bench { 1000.do { freqs.wchoose(weights.normalizeSum) } };
)

```

**Comment by smoge** [comment]
Adding `if (weights.sum == 1)` , when half of the lists are already normalized, give even worse performance than not checking at all.


One question: Using normalizeSum on a list with negative numbers is "legal" in supercollider? 

Consider this case:

```
[-1, -0.5, 0.5, 1].normalizeSum
[ -inf, -inf, inf, inf ]
```

**Comment by Asmatzaile** [comment]
I am trying to find a discussion around the original implementation of the `wchoose` and `windex` functions but it seems like they were there from [very early on](https://sourceforge.net/p/supercollider/supercollider/ci/e3a18fc9891d931f9b5b539758a7c6102d86184f/tree/build/SCClassLibrary/Common/Collections/SequenceableCollection.sc?force=True#l24).
FWIW, [some discussion about negative numbers and random.choice in python].

Are there any use cases where the user already has a normalized list without calling `.normalizeSum`? If not, I really don't see the value of keeping `wchoose` as it is. It's not a breaking change for older programs and a better API for newer ones. And why not, for users only wanting to normalize once and storing the normalized array for efficiency, adding an optional argument for not normalizing seems like a good idea.

**Comment by smoge** [comment]
> FWIW, [some discussion about negative numbers and random.choice in python]

I couldn't find the discussion (either sc or python). Where is the link?

I think numpy doesn't work well with negative numbers. 

`ValueError: negative dimensions are not allowed`
from: https://github.com/numpy/numpy/issues/13087


In SuperCollider we could allow but internally adjust to a positive range. I saw someone suggest this solution in a numpy discussion.

```c++
void adjust_to_positive(std::vector<double>& arr) {
  double min_value = *std::min_element(arr.begin(), arr.end());
  std::transform(arr.begin(), arr.end(), arr.begin(), [min_value](double val) {
    return val - min_value + 1;
  });
}
````

**Comment by telephon** [comment]
> Are there any use cases where the user already has a normalized list without calling `.normalizeSum`?

yes, it's the normal case, where you just write out a normalizes array, or you calculate it by hand and paste it in. This is at least what I have seen most, of course we don't know what is out there.

But while from the persective of practical use, I'd love to have a normalising wchoose, I still would prefer not to touch it and just add a letter to the end of the method name if I want it normalized. Then not everyone has to pay for a feature that is not needed in every case?

**Comment by telephon** [comment]
Just to be clear, we mean a method for those users who don't want the array normalized again for each new random value.

So the choice could be either:
1. wchoose is normalising wchoose1 non-normalising
2. wchoose is unchanged, the normalizing wchooseN is added

To decide, benchmarks would be needed, I think, because wchoose is in heavy use already.

**Comment by Asmatzaile** [comment]
How about adding optional arguments to `.choose`: (weights=nil, n_weights=nil)?
When n_weights!=nil, it could call to wchoose[nweights]; else if weights != nil, wchoose[weights.normalizeSum], and else, the usual `choose` function.
That would keep code that uses `wchoose` with the same good performance

**Comment by cdbzb** [comment]
I like this. One less method to have to remember.

**Comment by telephon** [comment]
Interesting idea, a little unusual, but possible.
You'd then write 
`.wchoose(n_weights:[1, 2, 3])`
instead of
`.wchooseN([1, 2, 3])`
right?

**Comment by Asmatzaile** [comment]
What I'm suggesting now is to keep `.wchoose` as is and change `.choose`.
`[/one, /two, /three].choose` would work like now.
But one could also do `[/one, /two, /three].choose([1,2,3])` for relative weights and the equivalent `[/one, /two, /three].choose([n_weights:[1/6,1/3,0.5])` for normalized weights.

**Comment by telephon** [comment]
ok, that is perhaps really the best solution. But then, why not just pass an array that will be normalised, and if you want to be efficient you use wchoose? Semantically slightly bent, but it is nozt uncommon that the most used method is easy but comes at a little cost.

**Comment by Asmatzaile** [comment]
Cool! Will make a PR later.
